### PR TITLE
Pose module unit tests

### PR DESCRIPTION
**Pull request by gmajrobotec**
# Description

## Abstract

This PR contains unit tests for the pose namespace, which is a part of traffic_simulator package

## Details
List of tests included in this PR:
1. pose.quietNaNPose
2. pose.quietNaNLaneletPose
3. pose.canonicalize_default
4. pose.canonicalize_invalid
5. pose.canonicalize_valid
6. pose.toMapPose_CanonicalizedLaneletPosepose.toMapPose_LaneletPose
7. pose.toCanonicalizedLaneletPose_noBoundingBox_noRoute_valid
8. pose.toCanonicalizedLaneletPose_noBoundingBox_noRoute_invalid
9. pose.toCanonicalizedLaneletPose_BoundingBox_noRoute_valid
10. pose.toCanonicalizedLaneletPose_BoundingBox_noRoute_invalid
11. pose.toCanonicalizedLaneletPose_BoundingBox_route_emptyInvalid
12. pose.toCanonicalizedLaneletPose_BoundingBox_route_emptyValid
13. pose.toCanonicalizedLaneletPose_BoundingBox_route_nonEmptyInvalid
14. pose.toCanonicalizedLaneletPose_BoundingBox_route_nonEmptyValid
15. pose.transformRelativePoseToGlobal
16. pose.relativePose_poses_zeros
17. pose.relativePose_poses_zero
18. pose.relativePose_poses_complex
19. pose.relativePose_first
20. pose.relativePose_second
21. pose.boundingBoxRelativePose_disjoint
22. pose.boundingBoxRelativePose_commonEdge
23. pose.boundingBoxRelativePose_intersect
24. pose.relativeLaneletPose_s_invalid
25. pose.relativeLaneletPose_s_valid
26. pose.relativeLaneletPose_offset_invalid
27. pose.relativeLaneletPose_offset_valid
28. pose.boundingBoxRelativeLaneletPose_s_invalid
29. pose.boundingBoxRelativeLaneletPose_s_valid
30. pose.boundingBoxRelativeLaneletPose_offset_invalid
31. pose.boundingBoxRelativeLaneletPose_offset_valid
32. pose.isInLanelet_inside
33. pose.isInLanelet_outsideFrontFar
34. pose.isInLanelet_outsideFrontClose
35. pose.isInLanelet_outsideBackFar
36. pose.isInLanelet_outsideBackClose
37. pose.isAtEndOfLanelets_noFollowing_within
38. pose.isAtEndOfLanelets_singleFollowing_within
39. pose.isAtEndOfLanelets_singleFollowing_outside
40. pose.isAtEndOfLanelets_multipleFollowing_within
41. pose.isAtEndOfLanelets_multipleFollowing_outside
42. pose.laneletLength_invalid
43. pose.laneletLength_valid

## References
Jira ticket: [internal link](https://tier4.atlassian.net/browse/RJD-1197)

# Destructive Changes

There are no destructive changes.

### CODE REVIEW

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+TEST_F(PoseTest, relativeLaneletPose_s_valid)
+{
+  const auto from =
+    traffic_simulator::helper::constructCanonicalizedLaneletPose(195, 0.0, 0.0, hdmap_utils);
+  const auto to =
+    traffic_simulator::helper::constructCanonicalizedLaneletPose(3002163, 0.0, 0.0, hdmap_utils);
+
+  const auto relative = traffic_simulator::pose::relativeLaneletPose(from, to, false, hdmap_utils);
+
+  EXPECT_NEAR(relative.s, 107.74, 0.001);
```
Comment (robomic) [inline simulation/traffic_simulator/test/src/utils/test_pose.cpp:443]: EXPECT_DOUBLE_NEAR

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+
+/**
+ * @note test calculation correctness when bounding boxes are disjoint
+ */
+TEST_F(PoseTest, boundingBoxRelativePose_disjoint)
+{
+  const auto pose_relative = makePose(makePoint(0.1108, -20.5955, 0.0));
+  const auto from = makePose(makePoint(81600.3967, 50094.4298, 34.6759));
+  const auto to = makePose(makePoint(81604.5076, 50071.8343, 40.8482));
+  const auto boundingBox = makeBoundingBox();
```
Comment (robomic) [inline simulation/traffic_simulator/test/src/utils/test_pose.cpp:377]: rename const auto bounding_box

**Review by robomic — CHANGES_REQUESTED** [review]
Only style changes, apply them throughout the entire file

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+        .longitude(138.8024583466017)
+        .altitude(0.0)))
+  {
+  }
+
+  std::shared_ptr<hdmap_utils::HdMapUtils> hdmap_utils;
+};
+
+/**
+ * @note test created object values
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:45]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+ */
+TEST(pose, quietNaNPose)
+{
+  const auto pose = traffic_simulator::pose::quietNaNPose();
+
+  EXPECT_TRUE(isnan(pose.position.x));
+  EXPECT_TRUE(isnan(pose.position.y));
+  EXPECT_TRUE(isnan(pose.position.z));
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:53]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+
+  EXPECT_EQ(pose.orientation.x, 0);
+  EXPECT_EQ(pose.orientation.y, 0);
+  EXPECT_EQ(pose.orientation.z, 0);
+  EXPECT_EQ(pose.orientation.w, 1);
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:58]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 6]
```diff
@@ -0,0 +1,669 @@
...
+TEST(pose, quietNaNLaneletPose)
+{
+  const auto pose = traffic_simulator::pose::quietNaNLaneletPose();
+
+  EXPECT_EQ(pose.lanelet_id, std::numeric_limits<std::int64_t>::max());
+  EXPECT_TRUE(isnan(pose.s));
+  EXPECT_TRUE(isnan(pose.offset));
+  EXPECT_TRUE(isnan(pose.rpy.x));
+  EXPECT_TRUE(isnan(pose.rpy.y));
+  EXPECT_TRUE(isnan(pose.rpy.z));
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:73]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+ * @note test conversion to geometry_msgs::msg::Pose from CanonicalizedLaneletPose
+ */
+TEST_F(PoseTest, toMapPose_CanonicalizedLaneletPose)
+{
+  const traffic_simulator::lanelet_pose::CanonicalizedLaneletPose canonicalized_pose(
+    traffic_simulator::helper::constructLaneletPose(195, 0.0, 0.0), hdmap_utils);
+
+  const geometry_msgs::msg::Pose pose = makePose(
+    makePoint(81585.1622, 50176.9202, 34.2595),
+    geometry_msgs::build<geometry_msgs::msg::Quaternion>().x(0).y(0).z(-0.6397).w(0.7686));
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:122]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 8]
```diff
@@ -0,0 +1,669 @@
...
+/**
+ * @note test function behavior with an empty unique_route_lanelets vector and a pose that can not be canonicalized
+ */
+TEST_F(PoseTest, toCanonicalizedLaneletPose_BoundingBox_route_emptyInvalid)
+{
+  const geometry_msgs::msg::Pose pose = makePose(makePoint(0, 0, 0));
+
+  EXPECT_EQ(
+    traffic_simulator::pose::toCanonicalizedLaneletPose(
+      pose, makeBoundingBox(), {}, true, 1.0, hdmap_utils),
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:217]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp (hunk 1 above)
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:443]: done

**Review by robomic — APPROVED** [review]
fix those style issues and it's good ;)

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 9]
```diff
@@ -0,0 +1,669 @@
...
+      pose, makeSmallBoundingBox(), true, 0.0, hdmap_utils),
+    std::nullopt);
+}
+
+/**
+ * @note Test function behavior with an empty unique_route_lanelets vector and a pose that can not be canonicalized.
+ */
+TEST_F(PoseTest, toCanonicalizedLaneletPose_BoundingBox_route_emptyInvalid)
+{
+  const geometry_msgs::msg::Pose pose = makePose(makePoint(0, 0, 0));
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:213]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 10]
```diff
@@ -0,0 +1,669 @@
...
+ */
+TEST_F(PoseTest, toCanonicalizedLaneletPose_BoundingBox_route_emptyValid)
+{
+  const auto lanelet_pose =
+    traffic_simulator::helper::constructCanonicalizedLaneletPose(195, 0.0, 0.0, hdmap_utils);
+
+  const geometry_msgs::msg::Pose pose = static_cast<geometry_msgs::msg::Pose>(lanelet_pose);
+
+  const auto canonicalized_pose = traffic_simulator::pose::toCanonicalizedLaneletPose(
+    pose, makeBoundingBox(), {}, true, 1.0, hdmap_utils);
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:232]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 11]
```diff
@@ -0,0 +1,669 @@
...
+    static_cast<traffic_simulator::LaneletPose>(canonicalized_pose.value()),
+    static_cast<traffic_simulator::LaneletPose>(lanelet_pose), 0.01);
+}
+
+/**
+ * @note Test function behavior with a non empty unique_route_lanelets vector and a pose that can not be canonicalized.
+ */
+TEST_F(PoseTest, toCanonicalizedLaneletPose_BoundingBox_route_nonEmptyInvalid)
+{
+  const geometry_msgs::msg::Pose pose = makePose(makePoint(0, 0, 0));
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:246]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 12]
```diff
@@ -0,0 +1,669 @@
...
+ */
+TEST_F(PoseTest, toCanonicalizedLaneletPose_BoundingBox_route_nonEmptyValid)
+{
+  const auto lanelet_pose =
+    traffic_simulator::helper::constructCanonicalizedLaneletPose(195, 0.0, 0.0, hdmap_utils);
+
+  const geometry_msgs::msg::Pose pose = static_cast<geometry_msgs::msg::Pose>(lanelet_pose);
+
+  const auto canonicalized_pose = traffic_simulator::pose::toCanonicalizedLaneletPose(
+    pose, makeBoundingBox(), {195}, true, 1.0, hdmap_utils);
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:265]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 13]
```diff
@@ -0,0 +1,669 @@
...
+
+/**
+ * @note Test calculation correctness when the pose lies outside the front of the lanelet, distance greater than the tolerance.
+ */
+TEST_F(PoseTest, isInLanelet_outsideFrontFar)
+{
+  const auto pose =
+    traffic_simulator::helper::constructCanonicalizedLaneletPose(3002163, -10.0, 0.0, hdmap_utils);
+
+  EXPECT_FALSE(traffic_simulator::pose::isInLanelet(pose, 3002163, 1, hdmap_utils));
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:563]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 14]
```diff
@@ -0,0 +1,669 @@
...
+
+/**
+ * @note Test calculation correctness when the pose lies outside the back of the lanelet, distance greater than the tolerance.
+ */
+TEST_F(PoseTest, isInLanelet_outsideBackFar)
+{
+  const auto pose =
+    traffic_simulator::helper::constructCanonicalizedLaneletPose(195, 120.0, 0.0, hdmap_utils);
+
+  EXPECT_FALSE(traffic_simulator::pose::isInLanelet(pose, 195, 2, hdmap_utils));
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:585]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 15]
```diff
@@ -0,0 +1,669 @@
...
+
+/**
+ * @note Test calculation correctness when the pose lies outside the back of the lanelet, distance smaller than the tolerance.
+ */
+TEST_F(PoseTest, isInLanelet_outsideBackClose)
+{
+  const auto pose =
+    traffic_simulator::helper::constructCanonicalizedLaneletPose(195, 110.0, 0.0, hdmap_utils);
+
+  EXPECT_TRUE(traffic_simulator::pose::isInLanelet(pose, 195, 10, hdmap_utils));
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp:596]: done

File: simulation/traffic_simulator/test/src/utils/test_pose.cpp [hunk 16]
```diff
@@ -0,0 +1,669 @@
...
+
+/**
+ * @note Test calculation correctness when the pose lies within the lanelet.
+ */
+TEST_F(PoseTest, isInLanelet_inside)
+{
+  const auto pose =
+    traffic_simulator::helper::constructCanonicalizedLaneletPose(195, 0.0, 0.0, hdmap_utils);
+
+  EXPECT_TRUE(traffic_simulator::pose::isInLanelet(pose, 195, 0, hdmap_utils));
```
[the remark this answers is not in the document]
Comment (gmajrobotec) [reply simulation/traffic_simulator/test/src/utils/test_pose.cpp]: changed to std::numeric_limits<double>::epsilon()

**Pull request merged**